contracts/token/ERC20/ERC20Permit.sol

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./ERC20.sol";
+import "./ERC20PermitStorage.sol";
+import "./IERC2612Permit.sol";
+
+// ERC20-Permit implementation from soliditylabs ( https://github.com/soliditylabs/ERC20-Permit )
+
+/**
+ * @dev Extension of {ERC20} that allows token holders to use their tokens
+ * without sending any transactions by setting {IERC20-allowance} with a
+ * signature using the {permit} method, and then spend them via
+ * {IERC20-transferFrom}.
+ *
+ * The {permit} signature mechanism conforms to the {IERC2612Permit} interface.
+ */
+abstract contract ERC20Permit is ERC20, IERC2612Permit {
+  constructor() {
```

Comment:
Avoid using constructors - instead put something in the test suite to validate that this is called in the end user's constructor or `initialize` function.  This is for proxy compatibility.